### DISCUSSION TITLE
Toggle between different chart type (e.g. candlestick chart)

### DISCUSSION THREAD

**Discussion by daniel-j-h** [tarkah/tickrs#30] [Ideas]
Hey hey :wave: I'm unsure about the charting library's limitations, but would it be possible to have e.g a candlestick plot type?

Should the user then be able to toggle between the charts and have a command line option to start with a specific chart type pre-selected? Just some ideas :hugs:

**Comment by tarkah** [comment]
Hmm, definitely not something supported out of the box but could potentially be hacked together. The library also provides a Canvas I'm sure it's possible to build my own using that. 

I like this idea as I like using candle charts as well, plus the API already gives me Open/Close/Low/High info. My biggest concern is guaranteeing all of the candlesticks will fit on the x axis depending on terminal size. So I either would have to implement a minimum screen width for that chart type and/or widen the intervals so there are less datapoints.

**Comment by tarkah** [comment]
@miraclx had a +1 for this request here:

https://github.com/tarkah/tickrs/issues/72#issuecomment-780696520

**Comment by miraclx** [comment]
🤷🏽‍♂

```text
─┬────┬───────┬───────┬────────┬───
 |                             |
 |   │     │ │                  
 |   │ │ │ │ │               │  
 |   █ █ │ █ █ │           │ │ │
 |   █ █ │ █ █ █       │ │ │ │ █
 |   █ │ █ █ █ █ │   │ █ █ █ █ █
 |   │ │ █ │ █ █ │ │ █ █ █ █ █ │
 |       │ │ █ │ █ │ █ █ █ █ │ │
 |           │ │ █ █ │ │ │ █ │  
 |             │ │ │ │ │ │ │    
 |                             | 
─┴────┴───────┴───────┴────────┴───
```

**Reply by miraclx** [reply]
yes it's hand-drawn 😪

**Reply by tarkah** [reply]
lol it's beautiful. Yeah I'm thinking 3 braille dots of width on the rectangles (Quartile 1 to Quartile 3) and a line from min to max along the middle braille dot of that rectangle. Then 1 dot of space. That means each candlestick will take up 4 dots, which is 2 text characters wide.

So the number we can fit is frame width / 2

**Reply by tarkah** [reply]
Disregard quartiles, I'm thinking of box/whisker. The rectangle will get drawn based on open of the first datapoint grouped for that candlestick, and the close of the last datapoint (updates with each new datapoint). High / Low is the max high of datapoints and min low of datapoints.

**Reply by miraclx** [reply]
3 braille dots for width, how about the height of the rectangles and can you show a preview of this?

**Reply by tarkah** [reply]
Height can be anything, depending on gap between open / close relative to the high / low of the day (y axis bounds). The number of braille dot available along the y axis is dependent on frame height. Each text "line" of height is 4 braille dots. 2x4 dots per text character.

I'll get a mockup done hopefully within the next day.

**Reply by tarkah** [reply]
Here's a mock-up... Haven't validated the accuracy of anything yet. It looks really cool! Need to add the lines next.

![image](https://user-images.githubusercontent.com/10239377/108280139-9ceccc00-7132-11eb-9cc0-8135897168b9.png)

**Reply by miraclx** [reply]
Okay that, for real, is dope as hell.... 🔥

Lemme jump on that PR and see it for myself 😅

**Reply by tarkah** [reply]
Ok, I'll push it shortly. It's SUPER unclean and I'm pretty sure it's not accurate, so I'll need to dig into it. But now that I can see it'll work, I'm very motivated :)

**Reply by tarkah** [reply]
High low lines!

![image](https://user-images.githubusercontent.com/10239377/108393042-1da9d780-71c8-11eb-8f29-7b36ad7c43e2.png)

**Reply by tarkah** [reply]
Coming along! Added y-labels and it also doesn't look half bad in summary mode

![image](https://user-images.githubusercontent.com/10239377/108400007-9bbdac80-71cf-11eb-8c6f-0e23a6177471.png)

**Reply by miraclx** [reply]
My terminal is all sorts of weird with braille in close proximity. Or maybe it's my fonts.

![Screenshot_20210218_233851](https://user-images.githubusercontent.com/16881812/108430987-8a61b800-7242-11eb-90c4-e5a46c498cbb.png)
![Screenshot_20210218_234107](https://user-images.githubusercontent.com/16881812/108431173-d3197100-7242-11eb-8285-e8a30464be14.png)

**Reply by tarkah** [reply]
Might be a combination of the two. I noticed the braille renders perfectly on alacritty / linux. But on windows no matter what I do, the rendering is only ok. Though nothing as bad as yours. The worst I get on windows (command line / powershell and even WSL from Windows Terminal) is a clear break between each row of braille, as seen here. 

![image](https://user-images.githubusercontent.com/10239377/108431864-53ff4b00-71f8-11eb-86bf-0ee6115f4e51.png)

Yours seem to be overlapping across the x-axis and causing issues that way. This is unfortunately just going to be a limitation of the braille characters used and how it's rendered on each font / platform / terminal.

**Reply by miraclx** [reply]
> testing with alacritty...

Okay I agree with you, it does look pretty good with that

![Screenshot_20210218_235514](https://user-images.githubusercontent.com/16881812/108432513-cbf36280-7244-11eb-86d1-e205b3cf7cc5.png)

![Screenshot_20210218_235537](https://user-images.githubusercontent.com/16881812/108432529-d57cca80-7244-11eb-8591-b1b39495d622.png)

**Reply by tarkah** [reply]
Nice! That looks good :D  

What do you think of the changes to the toggle block and help screen? I figured a wider toggle block is better than a taller one, plus it can be disabled anyways for those who are familiar with the controls.

Help was getting to tall so I made it wider instead.

**Reply by tarkah** [reply]
Alacritty on Windows with default Consolas font renders _almost_ perfectly too

![image](https://user-images.githubusercontent.com/10239377/108434856-2f59a200-71fd-11eb-914a-a85bde0b0e94.png)

**Reply by miraclx** [reply]
Yeah, I noticed the new look on the toggle block, just saw the help screen.

We can probably distribute items in the layouts more equally to minimize the height even further.

![Screenshot_20210219_002436](https://user-images.githubusercontent.com/16881812/108435010-e4657c00-7248-11eb-900b-a7d331f2017f.png)

**Reply by tarkah** [reply]
Good call. I just moved another item to the right column so the minimum height is a bit smaller now.

**Reply by miraclx** [reply]
So, I was able to resolve https://github.com/tarkah/tickrs/discussions/30#discussioncomment-382574 in my default non-alacritty terminal (Konsole) by disabling "Bi-Directional text rendering".

![Screenshot_20210220_183115](https://user-images.githubusercontent.com/16881812/108603720-e09d3b00-73a9-11eb-9229-893514c8ba7e.png)

**Comment by tarkah** [comment]
@daniel-j-h Candlestick charts have been implemented in the latest version! Check it out :D

**Reply by fspoettel** [reply]
Not the requester but I wanted to chime in and say that those look beautiful. Well done!

**Reply by tarkah** [reply]
Thanks, @fspoettel! I was pleasantly surprised with how well it turned out. It's definitely my preferred way to keep an eye on things.

I'm thinking of implementing a "full screen" mode to maximize room for the charting. And hopefully eventually adding some mouse interactivity, like a tool-tip on the Open / Close / High / Low data at the mouses x-axis location.

**Reply by daniel-j-h** [reply]
Awesome! :open_mouth: This is such a cool feature :clap: :clap: :clap:

**Comment by alfredo-catalano** [comment]
Kagi reversal charts based on high/low would be awesome. See => https://arpitnarechania.github.io/d3-kagi/

**Reply by tarkah** [reply]
Thanks for the suggestion! This would be really easy to implement now that I've already made a custom chart for candlesticks. I'll look into adding this sometime this week.

**Reply by tarkah** [reply]
I just read up on how Kagi works more in depth, so it won't be as easy as I was originally thinking. If I add this, I'd really like to try to make it accurate and I'll need to add some way to specify the "reversal amount" for each tab

**Reply by tarkah** [reply]
Got it figured out! Need to do some additional testing and add in some configuration. I'll give the ability to choose between high / low and close prices. And providing a Reversal Percentage vs Amount per ticker symbol. By default, I'll do 1% reversal for 1D and 4% reversal for all other timeframes.

![image](https://user-images.githubusercontent.com/10239377/108874402-f9c40880-75b0-11eb-9f14-e58093f41047.png)

Question... do you prefer that it stretches out to fill the entire x-axis when there aren't enough trend lines to fill the whole thing out? Or would you rather see it like this? And it may never fill out all the way just depending on # of reversals.

![image](https://user-images.githubusercontent.com/10239377/108875021-9e464a80-75b1-11eb-947f-fa6210a6002d.png)

**Reply by alfredo-catalano** [reply]
Damn, you're good! Can you choose a custom reversal amount? Nice job!

**Reply by tarkah** [reply]
Yeah, I'm working through that right now. At the very least, I'll add a section to the config file where you can specify the symbols and reversal amount  / pct and it'll use that. But I'd also like to add somewhere this can be customized while running the program. I have a few thoughts and will try to mock that up today.

**Reply by tarkah** [reply]
I don't think x-labels are going to work very reliably for this kind of chart, since time doesn't move across x-axis. So I may disable that and then wait to get mouse capture working so I can display the start / end date of each trend line in a tool tip or something.

**Comment by alfredo-catalano** [comment]
"Question... do you prefer that it stretches out to fill the entire x-axis when there aren't enough trend lines to fill the whole thing out? Or would you rather see it like this? And it may never fill out all the way just depending on # of reversals."

Whatever behavior adapts best to the current code?  My preference would be the latter.

**Reply by tarkah** [reply]
I agree the latter looks better.

Another thing I'm going to implement, is the ability to "scroll" horizontally across the x-axis for this chart type. In many cases, the number of reversals is greater than I can render based on the width of the terminal. So I'll always render it with the most recent number that can be rendered, then add the ability (and indicator), to scroll left (& back right) to show more datapoints. If you resize your terminal wider, they'd also show.

**Reply by tarkah** [reply]
And I'll see if I can get x-labels in a reasonable enough state before I ship this, because it may be a while before I get around to implementing any sort of mouse capture / tooltip

**Reply by alfredo-catalano** [reply]
Scrolling would be great. Something along the lines like this? => https://github.com/windelbouwman/lognplot

**Reply by tarkah** [reply]
I got scrolling working! I think it looks pretty great :) I used keybind <kbd>SHIFT</kbd>+<kbd><</kbd> &  <kbd>SHIFT</kbd>+<kbd>></kbd> to scroll left / right and the new indicator on the bottom right, similar to how it shows on the summary screen

![WzrbEGibTR](https://user-images.githubusercontent.com/10239377/109045026-f5b6ea00-7687-11eb-89d9-fa4410ca53cd.gif)

Last up is adding the ability to input custom reversal amounts via the gui and config file

**Reply by alfredo-catalano** [reply]
That is impressive!

**Reply by tarkah** [reply]
Thanks! I never even knew about Kagi until a few days ago. I really like this at the "1 week / 1 month" view to just keep an eye on trend changes.
